compiler/src/dmd/traits.d

```diff
@@ -739,6 +739,43 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
         auto se = new StringExp(e.loc, id.toString());
         return se.expressionSemantic(sc);
     }
+    if (e.ident == Id.fullyQualifiedName) // https://dlang.org/spec/traits.html#fullyQualifiedName
+    {
+        if (dim != 1)
+            return dimError(1);
+
+        Scope* sc2 = sc.push();
+        sc2.flags = sc.flags | SCOPE.noaccesscheck | SCOPE.ignoresymbolvisibility;
+        bool ok = TemplateInstance.semanticTiargs(e.loc, sc2, e.args, 1);
+        sc2.pop();
+        if (!ok)
+            return ErrorExp.get();
+
+        auto o = (*e.args)[0];
+        if (auto s = getDsymbolWithoutExpCtx(o))
+        {
+            if (s.semanticRun == PASS.initial)
+                s.dsymbolSemantic(null);
+
+            auto fqn = s.toPrettyChars().toDString();
+            assert(fqn);
+            auto se = new StringExp(e.loc, fqn);
+            return se.expressionSemantic(sc);
+        }
+        else if (auto t = getType(o))
+        {
+            auto fqn = t.toPrettyChars(true).toDString();
+            assert(fqn);
+            auto se = new StringExp(e.loc, fqn);
+            return se.expressionSemantic(sc);
```

Comment:
The if and else branches contain duplicated code. They could be rewritten to:

```d
string fqn;
if (auto s = getDsymbolWithoutExpCtx(o))
{
            if (s.semanticRun == PASS.initial)
                s.dsymbolSemantic(null);

            auto fqn = s.toPrettyChars().toDString();
}
else if (auto t = getType(o))
{
           auto fqn = t.toPrettyChars(true).toDString();
}
else { ... }

assert(fqn);
auto se = new StringExp(e.loc, fqn);
return se.expressionSemantic(sc);

```

Reply:
ok